### PR TITLE
Add autograd engine created from scratch

### PR DESCRIPTION
**Pull request by workofart**
## Summary
This PR adds the necessary building blocks for creating an autograd engine from scratch. For a demo on using it, there's an end-to-end training run in `tests/autograd/test_train.py` [link](https://github.com/workofart/ml-algorithms/blob/9f10bf33d2ea406a1f2895c18f1b3deb6d932246/test/autograd/test_train.py#L36-L55) for a binary classifier on the [breast cancer dataset](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_breast_cancer.html).

## Motivation
The primary motivation is to learn about neural networks from scratch and from first principles. All mathematical and calculus operations are explicitly derived in the code without abstraction, providing a comprehensive understanding. Additionally, debugging a neural network, especially the `backward()` implementations of various functions (unary, arithmetic, loss, and activation), offers a rewarding learning experience.

## Inspiration
This implementation took inspiration from [Micrograd](https://github.com/karpathy/micrograd/tree/master), and kept the API interface as close as possible to [Pytorch](https://github.com/pytorch/pytorch/tree/main).

## Details
More specifically, this includes:
- tensor
  - add, mul, matmul, pow, sub, division, neg ops)
  - forward, backward, reshape
- nn
  - Module (base class)
  - Linear (basic building block for a perceptron/hidden layer)
- functional (including backprop derivation defined in `backward()`)
  - relu (activation function)
  - sigmoid (activation function)
  - binary_cross_entropy (loss)
- optim
  - Optimizer (base class)
  - SGD (basic stochastic gradient descent optimizer)
- test/
  - All the unit tests for the above functionality
  - End-to-end training run by combining all the components on a real dataset

## Test Cases

image

## Next Steps
Consider expanding the engine's functionality to eventually incorporate all state-of-the-art neural network architectures. While efficient training is not guaranteed, the primary purpose is for learning and experimentation.

### CODE REVIEW

**Comment by workofart** [comment]
Moved the code to a separate repository: https://github.com/workofart/ml-by-hand
